File: 2262-total-appeal-of-a-string/2262-total-appeal-of-a-string.py
class Solution:
    def appealSum(self, s: str) -> int:
        occur_pos = {}
        for c in string.ascii_lowercase:
            occur_pos[c] = -1
        prev = [0] * len(s)
        dp = [0] * len(s)
        for i in range(len(s)):
            prev[i] = (prev[i-1] if i>0 else 0) + (i - occur_pos[s[i]])
            dp[i] = (dp[i-1] if i>0 else 0) + prev[i]
            occur_pos[s[i]] = i
        return dp[len(s)-1]

File: 2262-total-appeal-of-a-string/2262-total-appeal-of-a-string.cpp
class Solution {
public:
    long long appealSum(string s) {
        unordered_map<char,int> prev;
        vector<int> dp(2);
        dp[0] = 1;
        long long res = 1;
        prev[s[0]] = 0;
        for (int i = 1; i < s.size(); i++) {
            if (prev.find(s[i]) == prev.end()) {
                prev[s[i]] = -1;
            }
            dp[i%2] = dp[(i-1)%2] + i - prev[s[i]];
            prev[s[i]] = i;
            res += dp[i%2];
        }
        return res;
    }
};